blockchain/upgrade.go

```diff
@@ -107,41 +139,60 @@ func readBlockTree(v1BlockIdxBucket database.Bucket) (map[chainhash.Hash][]*chai
 		}
 
 		blockHash := header.BlockHash()
-		childBlocksMap[header.PrevBlock] =
-			append(childBlocksMap[header.PrevBlock], &blockHash)
+		prevHash := header.PrevBlock
+
+		if blocksMap[blockHash] == nil {
+			blocksMap[blockHash] = &blockChainContext{height: -1}
+		}
+		if blocksMap[prevHash] == nil {
+			blocksMap[prevHash] = &blockChainContext{height: -1}
+		}
+
+		blocksMap[blockHash].parent = &prevHash
+		blocksMap[prevHash].children =
+			append(blocksMap[prevHash].children, &blockHash)
 		return nil
 	})
-	return childBlocksMap, err
+	return blocksMap, err
 }
 
 // determineBlockHeights takes a map of block hashes to a slice of child hashes
 // and uses it to compute the height for each block. The function assigns a
 // height of 0 to the genesis hash and explores the tree of blocks
 // breadth-first, assigning a height to every block with a path back to the
-// genesis block.
-func determineBlockHeights(childBlocksMap map[chainhash.Hash][]*chainhash.Hash) map[chainhash.Hash]uint32 {
-	blockHeights := make(map[chainhash.Hash]uint32)
+// genesis block. This function modifies the height field on the blocksMap
+// entries.
+func determineBlockHeights(blocksMap map[chainhash.Hash]*blockChainContext) {
 	queue := list.New()
 
-	// The genesis block is included in childBlocksMap as a child of the zero
-	// hash because that is the value of the PrevBlock field in the genesis
-	// header.
-	for _, genesisHash := range childBlocksMap[zeroHash] {
-		blockHeights[*genesisHash] = 0
+	// The genesis block is included in blocksMap as a child of the zero hash
+	// because that is the value of the PrevBlock field in the genesis header.
+	for _, genesisHash := range blocksMap[zeroHash].children {
```

Comment:
Same here regarding panic if the hash doesn't exist since it will attempt to dereference nil.  It previously wouldn't panic because the range was over the entry and range over nil produce no crash or iterations.

There are several more instances of this below as well.

I will concede that it _shouldn't_ happen that these hashes don't exist because it was built from a supposedly valid chain, so it might not be worth worrying about.

Reply:
Added a check for this case.